### PR TITLE
Nucular Fishin' Engine

### PR DESCRIPTION
**Pull request by amylizzle**
[INPUT WANTED][FEATURE]
## About the PR 

🐟 

Status: APPROVED :dancers: 

Adds a nuclear fission engine with matsci and gas integration to the available power generating engines.

Design doc: https://hackmd.io/@amylizzle/nuclearfission

Draft wiki page: https://wiki.ss13.co/User_talk:Amylizzle
image

## Why's this needed? 
More engine good.

## Changelog
```changelog
(u)Amylizzle, Danger Noodle, WALPVRGIS, ZeWaka
(*)Adds a nuclear fission engine which replaces the TEG on Clarion. Sprites by WALPVRGIS and ZeWaka, mapping by Danger Noodle.
(+)Make reactor components out of mined and alloyed material: pizza control rods, ice heat exchangers, erebite fuel rods.
(+)Feed gas into the reactor to cool it and use the heat to generate power in the turbine.
(+)Mechcomp integration as standard!
```

### CODE REVIEW

**Comment by Josephin23** [comment]
"matsci integration" yup, can already tell this will go very badly
seriously though i'm always up for more power generation methods, TEG gets stale

**Comment by amylizzle** [comment]
!merge_upstream

**Comment by notfrying1pans** [comment]
is this the 4th or 5th nuclear engine made for goon

**Comment by amylizzle** [comment]
!merge_upstream

**Comment by amylizzle** [comment]
!merge_upstream

**Comment by amylizzle** [comment]
!merge_upstream

File: code/procs/scanprocs.dm
Comment (ZeWaka) [inline code/procs/scanprocs.dm:201]: R?.stage moment

Comment (amylizzle) [reply code/procs/scanprocs.dm:201]: `if(null && null.property)` won't runtime because the first check evaluates false and the second check is thus never run

Comment (amylizzle) [reply code/procs/scanprocs.dm:201]: oh wait, you were saying I could replace this whole thing with `R?.stage`. I am dum.

**Comment by ThatFiveGuys** [comment]
I still have no clue what making full rad immunity not full rad immunity will add to the game outside of annoying miners and those with protection they earned chilling in radstorms microdoses of tox which seems overall anti qol for no reason. If you want to have dealing with high grade radiation fuel dangerous, split up neutron and gamma radiation resistance and never have any items with full neutron resistance, you maintain the qol and you get to have some fuels be more dangerous to deal with. Similar thoughts for rad resistance gene. Making all radiation deal burn, tox and organ damage seems needlessly cruel, ruins the niche of neutronium doing tox + burn, and more importantly the organ damage is hell to heal and radiation tends to get dealt in mass when it comes to radstorms and the nuclear engine, while organ healing requires very slow and personalized healing.

**Comment by ThatFiveGuys** [comment]
While visible radiation will definitely help early stage mining, my point isnt that. My point is that what does making miners and roaches have to heal 10 tox every time they want to use the items they worked for to wait out a radstorm add to the game. Being able to ignore radstorms for example, very much adds to the game. Not only does it make certain genes have a strong selling factor to account for there cons (empowered radres costs stability and roaches get far more brute), it also allows you to engage with these elements far more differently, with you being free to break into places during radstorms, help those dying who couldnt get to maint, etc. I do very much agree that meltdowns should effect everyone, however i think theres a clear and better solution for it. Engineers should get hurt far more by meltdowns then the crew, and engineers shouldnt be able to ignore the hurting. By seperating neutron and gamma radiation resistance, having no such thing as full neutron resistance, and having meltdowns produce neutron radiation thats roughly the size of the engine room and normal radiation spew everywhere, you have meltdowns fuck up engineers, you give the crew good counterplay to engineers fuckups by snagging rad suits and the like, and you also you keep the previous qol/gameplay aspects.

**Comment by amylizzle** [comment]
Again, I counsel actually playing it before you worry too much. A rad suit is plenty to tank being stood in a radiation blowout event. It really isn't a huge difference - what you were immune to before, you are are 99% immune to now. The only time this is really going to come into play for a miner would be if you collected all the radioactive ore you'd collected in a pile and then rolled around in it. 
The lack of immunity applies in extreme situations, not in common ones. Don't be an idiot with radioactive things and you won't even notice the difference. 

Also if it does turn out to suck for whatever reason, it's literally a one line change.

**Comment by ThatFiveGuys** [comment]
The point is giving negligible amounts of tox to people adds nothing to the game, having a minimal amount of damage radiation has to do in order for it to actually damage seems good, that way you dont give people annoyingly low amounts of tox.

**Comment by amylizzle** [comment]
You don't start to take tox damage immediately, until you reach a total dose of 0.4 Sv (ie, you've been holding a neutronium fuel rod for 10 minutes).

**Comment by ThatFiveGuys** [comment]
I dont think you should take damage whatsoever from radstorms with full radiation protection and im unsure if thats a thing with how its implemented currently

**Comment by nefarious6th** [comment]
maybe split the radiation re-do into a separate PR to be patched for approval/feedback first and then leave the fission engine wrapped up in this PR? lot of change going on at once and the agreement/disagreement with radiation changes is something that while wrapped up on this PR i think has the ability to cause some debate while not necessarily having any real bearing on the engine you're trying to make here (does a winded debate on what the merits and unique flavor of roachpeople should be necessarily have to be tacked onto a new engine that's utilizing radioactive materials, etc.)

**Comment by amylizzle** [comment]
!merge_upstream

**Comment by amylizzle** [comment]
!merge_upstream

**Comment by amylizzle** [comment]
DNM until I undo the changes to _build.dm that I was using for testing

File: _std/__build.dm
Comment (ZeWaka) [inline _std/__build.dm:34]: if you're going to undraft this you gotta fix this stuff

**Review by ZeWaka — CHANGES_REQUESTED** [review]

File: code/mob.dm
Comment (Tarmunora) [inline code/mob.dm:3143]: weird scaling on the rad protection there, especially since all existing sources are assuming a 0-100(+) percentage scale.

File: code/mob.dm
Comment (Tarmunora) [inline code/mob.dm:225]: the decay rate seems comically small for the average round length. If you take a 0.4sv dose (the upper end of the lowest damage scale for rads), you will take damage 200 times before the rads decay to safe levels (taking ~140 tox damage over those 12 minutes or so minutes) barring medical intervention.

File: code/datums/components/radioactive.dm
Comment (Tarmunora) [inline code/datums/components/radioactive.dm:108]: Why is it the same dose from touching something, standing nearby it for a second, and *eating it*?

File: code/datums/components/radioactive.dm
Comment (Tarmunora) [inline code/datums/components/radioactive.dm:104]: Should probably put a short cooldown on this, unless potential doubledipping the proc on stuff like COMSIG_ATTACKHAND, COMSIG_ITEM_PICKUP is intentional.

**Review by Tarmunora — CHANGES_REQUESTED** [review]
reviewed up to [code/mob/living/object.dm](https://github.com/goonstation/goonstation/pull/8767/files#diff-cbaf8d712b06e62a10c559b8d38501d77bad1be5f9ec63eea93b2e573a06675d) (i.e.: just radiation rework stuff, and not all of it)

Overall, I think the new radiation system is probably too nasty. While I haven't gotten to looking much at radiation *sources*, I think that as it stands, the damage scaling at the upper end is probably too high for a linear dosage scale, especially since the decay rate is so low as to be effectively non-existent once you're irradiated enough to be in the second damaging tier of the scale.

File: code/modules/status_system/statusEffects.dm
Comment (amylizzle) [reply code/modules/status_system/statusEffects.dm:450]: intentional. Suddenly vomiting can be stunning. I can change if needed.

File: code/modules/status_system/statusEffects.dm
Comment (amylizzle) [reply code/modules/status_system/statusEffects.dm:407]: Fair, maybe stages should be exponential instead of linear? I'll play with it

File: code/datums/abilities/werewolf.dm
Comment (amylizzle) [reply code/datums/abilities/werewolf.dm:74]: that was actually supposed to be both. Oversight on my part.

File: code/datums/components/radioactive.dm
Comment (amylizzle) [reply code/datums/components/radioactive.dm:108]: Because I copy-pasted my code and forgot to change the values.

File: code/datums/components/radioactive.dm
Comment (amylizzle) [reply code/datums/components/radioactive.dm:104]: A little bit of double dipping is intentional, because standing next to multiple rad sources should be more deadly than standing next to one. I'll put a short (sub item tick time) cooldown on it anyway though, to stop something like dumping a stack of ore next to someone being instantly fatal.

File: code/datums/components/radioactive.dm
Comment (amylizzle) [reply code/datums/components/radioactive.dm:48]: essentially because `COMSIG_ITEM_PROCESS` only gets sent to items registered in `global.processing_items`, and those items need to implement `process()` anyway. I suppose I could register the `COMSIG_ITEM_PROCESS` signal and leave the `process()` as a stub instead, but it's functionally equivalent.

File: code/datums/components/radioactive.dm
Comment (amylizzle) [reply code/datums/components/radioactive.dm]: fixed

File: code/modules/status_system/statusEffects.dm
Comment (amylizzle) [reply code/modules/status_system/statusEffects.dm:407]: I might have gone a little too far the other way, but see what you think

**Comment by amylizzle** [comment]
I split this off into #9741 - once that's closed this should be like <10 files

**Comment by amylizzle** [comment]
!merge_upstream

**Comment by amylizzle** [comment]
!merge_upstream

**Comment by amylizzle** [comment]
!merge_upstream

**Comment by amylizzle** [comment]
!merge_upstream

**Comment by amylizzle** [comment]
!merge_upstream

**Comment by amylizzle** [comment]
!merge_upstream

**Comment by Azrun** [comment]
Where do you get the fish? Where do you put the fish?

**Comment by amylizzle** [comment]
!merge_upstream

**Comment by amylizzle** [comment]
!merge_upstream

File: code/obj/nuclearreactor/reactorcomponents.dm
Comment (Sovexe) [inline code/obj/nuclearreactor/reactorcomponents.dm:5]: Missing /item/ ?

Comment (amylizzle) [reply code/obj/nuclearreactor/reactorcomponents.dm:5]: oops

**Comment by ZeWaka** [comment]
Hate to be the one to say this, but this is far from final review. It needs a lot more spritework before merging (I say this as the person who has been very very slowly working on the reactor sprite that /anyone/ is welcome to pick up and work on).

**Comment by amylizzle** [comment]
a *lot* more spritework? Aside from the reactor sprite, what still needs doing? 

I only ask because I've basically just been trying to keep the branch in a mergeable state while waiting for review, I didn't really think anything more needed to be done.

**Review by Azrun — COMMENTED** [review]
Uh, so I started a review...

File: code/obj/nuclearreactor/centrifuge.dm
Comment (ZeWaka) [inline code/obj/nuclearreactor/centrifuge.dm:19]:
```suggestion
	var/static/sound/sound_load = sound('sound/items/Deconstruct.ogg')
	var/static/sound/sound_process = sound('sound/effects/pop.ogg')
	var/static/sound/sound_grump = sound('sound/machines/buzz-two.ogg')
```

File: code/obj/nuclearreactor/centrifuge.dm
Comment (ZeWaka) [inline code/obj/nuclearreactor/centrifuge.dm:54]: does this properly check interact range?

File: code/obj/nuclearreactor/centrifuge.dm
Comment (ZeWaka) [inline code/obj/nuclearreactor/centrifuge.dm:58]:
```suggestion
			user?.u_equip(W)
```

File: code/obj/nuclearreactor/nuclearreactor.dm
Comment (ZeWaka) [inline code/obj/nuclearreactor/nuclearreactor.dm:14]: ?

File: code/obj/nuclearreactor/nuclearreactor.dm
Comment (ZeWaka) [inline code/obj/nuclearreactor/nuclearreactor.dm:3]:
```suggestion
// ----------------------------------------------------- //
// Defintion for the nuclear reactor engine
// ----------------------------------------------------- //
```
fucks up code docs otherwise

File: code/obj/nuclearreactor/nuclearreactor.dm
Comment (ZeWaka) [inline code/obj/nuclearreactor/nuclearreactor.dm:279]: :)

File: code/obj/nuclearreactor/nuclearreactor.dm
Comment (ZeWaka) [inline code/obj/nuclearreactor/nuclearreactor.dm:313]: ?

**Review by ZeWaka — CHANGES_REQUESTED** [review]

File: code/obj/nuclearreactor/turbine.dm
Comment (ZeWaka) [inline code/obj/nuclearreactor/turbine.dm:39]:
```suggestion
	var/static/list/grump_sound_list = list('sound/machines/engine_grump1.ogg','sound/machines/engine_grump2.ogg','sound/machines/engine_grump3.ogg', 'sound/impact_sounds/Metal_Clang_1.ogg', 'sound/impact_sounds/Metal_Hit_Heavy_1.ogg')
```

File: code/obj/nuclearreactor/turbine.dm
Comment (ZeWaka) [inline code/obj/nuclearreactor/turbine.dm:32]:
```suggestion
	var/static/sound_stall = 'sound/machines/tractor_running.ogg'
```

**Review by ZeWaka — CHANGES_REQUESTED** [review]
this too

File: code/obj/nuclearreactor/centrifuge.dm
Comment (amylizzle) [reply code/obj/nuclearreactor/centrifuge.dm:54]: If it doesn't, a lot of other things are broken. I'm pretty sure it does.

File: code/obj/nuclearreactor/turbine.dm
Comment (ZeWaka) [inline code/obj/nuclearreactor/turbine.dm:90]: commented out code?

File: code/obj/nuclearreactor/turbine.dm
Comment (ZeWaka) [inline code/obj/nuclearreactor/turbine.dm:40]: code doc comments

File: code/obj/nuclearreactor/turbine.dm
Comment (ZeWaka) [inline code/obj/nuclearreactor/turbine.dm:11]: TODO as in this PR?

File: code/obj/nuclearreactor/reactorcontrol.dm
Comment (ZeWaka) [inline code/obj/nuclearreactor/reactorcontrol.dm:149]: you've checked the performance of this, right?

File: code/obj/nuclearreactor/reactorcomponents.dm
Comment (ZeWaka) [inline code/obj/nuclearreactor/reactorcomponents.dm:227]: code doc comment

File: code/obj/nuclearreactor/reactorcomponents.dm
Comment (ZeWaka) [inline code/obj/nuclearreactor/reactorcomponents.dm:133]: isn't this list only supposed to have neutrons in it?

File: code/obj/nuclearreactor/reactorcomponents.dm
Comment (ZeWaka) [inline code/obj/nuclearreactor/reactorcomponents.dm:99]: ditto but components

File: code/obj/nuclearreactor/turbine.dm
Comment (amylizzle) [reply code/obj/nuclearreactor/turbine.dm:11]: Nah, this was an early thing that has since been abandoned. will remove.

File: code/obj/nuclearreactor/reactorcontrol.dm
Comment (amylizzle) [reply code/obj/nuclearreactor/reactorcontrol.dm:149]: I have, it's not bad. It's only 49 ops. The real performance issue is the neutrons, which is why the reactor explodes when the radiation level goes over a threshold, since it indicates an exponential process of generating more neutrons.

Profiling wise though, the reactor is surprisingly low cost except when it's right about to explode, and that state only lasts a couple machine ticks.

**Review by ZeWaka — APPROVED** [review]

**Comment by amylizzle** [comment]
!merge_upstream

**Pull request merged**
